### PR TITLE
Use Lua script to manage multi-language redirects

### PR DESCRIPTION
**Pull request by enricostano**
### WAT

One of the reasons we want to migrate our static pages to our own VPS is to have more control on how requests are managed. In this case we want to redirect requests to our (future) `/es`, `/ca` and `/en` URLs in order to show the correct language based on the `Accept-Language` header set by the browser.

### CODE REVIEW

**Comment by rewritten** [comment]
Interesting, but why doing it for each request when you can just generate an nginx configuration with the supported languages? I mean, iterating languages in a jinja template for the nginx configuration.

**Comment by enricostano** [comment]
@rewritten and what about default redirect?

**Comment by rewritten** [comment]
From the page https://www.nginx.com/resources/wiki/modules/accept_language/ there is a very simple nginx-only alternative:

```
# accept-language: en,en-US;q=0.8,ja;q=0.6
set $first_language $http_accept_language;
if ($http_accept_language ~* '^(.+?),') {
    set $first_language $1;
}

set $language_suffix 'en';
if ($first_language ~* 'ja') {
    set $language_suffix 'ja';
}

# repeat the last `if` for each supported language
```

Maybe too much use of `if`, but still why pulling in the extras?

**Comment by enricostano** [comment]
Really interesting, although compiling an unofficial and quite dated (2013) module is a bit scary to me. The `nginx-extras` package doesn't seem that bad to me.

EDIT: and still, you need to do that for each request, right?

**Comment by rewritten** [comment]
I wasn't implying compiling an unofficial plugin, but rather adding the `if`-based configuration snippet plainly in the configuration files. It's not comparable with running LUA, because the configuration is loaded, parsed and transformed into a FSM at load time, and then run super-fast inside the main nginx process. On the other side, LUA needs at least one context-switch per call.

It could be interesting if you want to add cookies to the responses etc, but in that specific case, why forcing esoteric languages into a piece of configuration we don't control much, when there are plenty of very lightweight tools in ruby or javascript which can be put behind nginx, serve static pages and add all the redirects, cookies etc that you can imagine?

In other words, either we want static pages, or we don't want static pages. If we need more than static pages, my opinion is that we should try to use tools that we already control, and not new languages, or plug-ins to pieces of infrastructure that are usually a given. The same would apply to using LUA in postgresql. Unless there is a very compelling reason to do it, why bother?

**Comment by enricostano** [comment]
Well, having 1 web server is better than having 2 IMHO.

I would prefer to not add new pieces and use plain NGINX and it seems that the `if` based solution could be a nice one. I'll give it a try and see if it covers our needs. Thanks for the inputs!

EDIT: well maybe the `if` way is not our best option https://www.nginx.com/resources/wiki/start/topics/depth/ifisevil/

**Comment by fcsonline** [comment]
:+1:

**Comment by fcsonline** [comment]
:clap:
